# MINI PROJECT - Day3 & Day4
# (Titanic Data Analysis)

## TASK 1 --Import

In [2]:
import pandas as pd   #import lib

In [3]:
tit = pd.read_csv('train.csv')   #load dataset

In [4]:
tit

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


## TASK 2 -- Dataset Overview

In [5]:
tit.head()   #first 5 rows

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [6]:
  tit.tail()  #last 5 rows

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.00,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.00,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.45,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.00,C148,C
890,891,0,3,"Dooley, Mr. Patrick",male,32.0,0,0,370376,7.75,NaN,Q


In [7]:
tit.shape 

(891, 12)

In [8]:
tit.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [9]:
tit.info()  
# type(data.info())   #data types of info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [10]:
tit.describe()   #summary statistics

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


### Observations

- The dataset contains passenger information from the Titanic.
- There are both numerical and categorical columns.
- The target variable is `Survived`.
- Some columns have missing values (confirmed using `info()`).

## TASK 3  -- Data Cleaning

In [11]:
#tit.isna()       #check missing values-(NaN, None, NaT)

tit.isna().sum()  #This shrinks those 891 rows of confusing booleans into a single, clean list of numbers.


PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [12]:
tit[tit['Age'].isna()]     #this can give the exact no of rows that are broken

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
17,18,1,2,"Williams, Mr. Charles Eugene",male,NaN,0,0,244373,13.0000,NaN,S
19,20,1,3,"Masselmani, Mrs. Fatima",female,NaN,0,0,2649,7.2250,NaN,C
26,27,0,3,"Emir, Mr. Farred Chehab",male,NaN,0,0,2631,7.2250,NaN,C
28,29,1,3,"O'Dwyer, Miss. Ellen ""Nellie""",female,NaN,0,0,330959,7.8792,NaN,Q
...,...,...,...,...,...,...,...,...,...,...,...,...
859,860,0,3,"Razi, Mr. Raihed",male,NaN,0,0,2629,7.2292,NaN,C
863,864,0,3,"Sage, Miss. Dorothy Edith ""Dolly""",female,NaN,8,2,CA. 2343,69.5500,NaN,S
868,869,0,3,"van Melkebeke, Mr. Philemon",male,NaN,0,0,345777,9.5000,NaN,S
878,879,0,3,"Laleff, Mr. Kristo",male,NaN,0,0,349217,7.8958,NaN,S


In [19]:
# tit.dropna()    # remove missing values -- this is a bad idea this can loose ur entire datset

tit['Cabin'] = tit['Cabin'].fillna('unknown')   #hnadling cabin column

median_age = tit['Age'].median()              # handle age column
tit['Age'] = tit['Age'].fillna(median_age) 

tit.dropna(subset=['Embarked'], inplace =True)   #hnadling embarked column

tit.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

In [18]:
# CHECKING DUPLICATES

# tit.duplicated()             # --detecting
tit.duplicated().sum()     #-- how many duplicates

np.int64(0)

### Observation
* Age has 177 missing values.
* Cabin has 687 missing values.
* Embarked has 2 missing values.

## TASK 4  -- Data Analysis

In [16]:
# HOW MANY PASSENGERS SURVIVED?
tit['Survived'].sum()

np.int64(340)

In [25]:
# HOW MANY PASSENGERS DID NOT SURVIVE?
(tit['Survived'] == 0).sum()

np.int64(549)

In [75]:
# How many male and female passengers were there?
# tit['Sex'].count()
tit['Sex'].value_counts()  # value_counts this is good for categorical data 

Sex
male      577
female    312
Name: count, dtype: int64

In [39]:
# How many passengers belonged to:
# First Class
# Second Class
# Third Class

tit['Pclass'].value_counts()

Pclass
3    491
1    214
2    184
Name: count, dtype: int64

In [40]:
#Average age of passengers.
tit['Age'].mean()

np.float64(29.315151856017994)

In [42]:
# Average fare paid.
tit['Fare'].mean()

np.float64(32.09668087739032)

In [67]:
# Which gender had a higher survival count?

 ##---- To analyze two columns together, use double square brackets [[ 'Col1', 'Col2' ]] to select them as a new DataFrame

tit.groupby('Sex')['Survived'].sum()     #in ()bracket we add that column jis category ko lekr pucha gyauska col, nd []bracket mei woh jo pucha gya use related column 
# ans is female

Sex
female    231
male      109
Name: Survived, dtype: int64

In [69]:
# Which passenger class had the highest survival count?
tit.groupby('Pclass')['Survived'].sum()
# ans is 1st class

Pclass
1    134
2     87
3    119
Name: Survived, dtype: int64

In [72]:
# Passengers older than 60.
# How many?

(tit['Age']>60).sum()

np.int64(21)

In [73]:
# Passengers younger than 18.
# How many?

(tit['Age']<18).sum()

np.int64(113)

### Observation
* Only 340 passengers survived out of 891.
* Number of male were more than female .
* Majority of the passenger are from third class.
* Average age of passenger was approximately 29.
* Female passengers survived more often than male.
* The passengers belong to 1st class had highest survival count.
* There were 113 children and 21 senior citizen.
  

### Research
#### Why is Pandas important for Data Analysis?
Pandas is important because it turns slow, messy data tasks into fast, automated code. It saves time, prevents mistakes, and gives you total control over your data.

#### Why companies use Pandas.
* Saves Money: Automates boring, repetitive manual reporting.
* Builds Trust: Code prevents human copy-paste errors.
* Saves Knowledge: Scripts show exactly how metrics were calculated

#### Where Data Analysts use it.
* Cleaning Data: Fixing messy text, missing numbers, and duplicates.
* Exploring Trends: Finding quick averages, patterns, and anomalies.
* Merging Sources: Combining different files into one final report.

#### Advantages over Excel.
* Never Crashes: Handles millions of rows without freezing.
* Easy Auditing: Shows step-by-step logic instead of hidden cell formulas.
* One-Click Run: Re-runs weekly reports instantly with one button.

#### Why it is useful for large datasets.
* Blazing Fast: Uses optimized speed to calculate columns simultaneously.
* Lightweight Memory: Compresses huge files to fit on standard laptops.
* No Row Limits: Breaks past Excel's strict 1-million-row ceiling.